cellxgene_schema_cli/tests/test_validate.py

```diff
@@ -0,0 +1,408 @@
+        self.schema_def = validate._get_schema_definition(examples.SCHEMA_VERSION)
+        self.good_uns = examples.good_uns
+        self.bad_uns = examples.bad_uns
+
+    def test_validate_good_uns(self):
+        self.validator._validate_dict(
+            self.good_uns, "uns", self.schema_def["components"]["uns"]
+        )
+        self.assertFalse(self.validator.errors)
+
+    def test_validate_bad_uns(self):
+
+        # Do one key at a time but skip schema_version as that's not checked here
+        for key, value in self.bad_uns.items():
+
+            if key == "schema_version":
+                continue
+
+            current_dict = self.good_uns.copy()
+            current_dict[key] = value
+
+            self.validator.errors = []
+            self.validator._validate_dict(
+                current_dict, "uns", self.schema_def["components"]["uns"]
+            )
+            self.assertTrue(self.validator.errors)
+
+
+class TestColumnValidation(unittest.TestCase):
+    def setUp(self):
+        self.validator = validate.Validator()
+        self.validator.adata = examples.adata_empty
+        self.good_var = examples.good_var
+        self.good_obs = examples.good_obs
+        self.bad_var = examples.bad_var
+        self.bad_obs = examples.bad_obs
+        self.schema_def = validate._get_schema_definition(examples.SCHEMA_VERSION)
+
+        self.unique = pd.DataFrame(
+            [["abc", "def"], ["ghi", "jkl"], ["mnop", "qrs"]],
+            index=["X", "Y", "Z"],
+            columns=["col1", "col2"],
+        )
+
+        self.duped = pd.DataFrame(
+            [["abc", "def"], ["ghi", "qrs"], ["abc", "qrs"]],
+            index=["X", "Y", "X"],
+            columns=["col1", "col2"],
+        )
+
+        self.column_def_uniq = {"unique": True}
+        self.column_def_not_uniq = {"unique": False}
+
+    def test_validate_unique_good(self):
+
+        self.validator._validate_column(
+            self.unique.index, "index", "unique_df", self.column_def_uniq
+        )
+        self.assertFalse(self.validator.errors)
+
+        self.validator._validate_column(
+            self.unique["col1"], "col1", "unique_df", self.column_def_uniq
+        )
+        self.assertFalse(self.validator.errors)
+
+    def test_validate_unique_valid_duped(self):
+
+        self.validator._validate_column(
+            self.duped["col1"], "col1", "duped_df", self.column_def_not_uniq
+        )
+        self.assertFalse(self.validator.errors)
+
+    def test_validate_unique_invalid_duped(self):
+
+        self.validator._validate_column(
+            self.duped.index, "index", "duped_df", self.column_def_uniq
+        )
+        self.assertTrue(self.validator.errors)
+
+    def test_validate_unique_invalid_duped2(self):
+
+        self.validator._validate_column(
+            self.duped["col1"], "col1", "duped_df", self.column_def_uniq
+        )
+        self.assertTrue(self.validator.errors)
+
+    def test_ontology_columns(self):
```

Comment:
This test is a schema-level test, while the ones above seem to be lower-level column validation logic tests that are not tied to a schema. Thinking it might good (throughout all tests) to separate those that are testing schema constraints vs the lower level implementation of schema-agnostic constraints.  Could even separate into different test files, so it's clear that, say, one file is all about testing schema 2.0.0 specifically.